internal/supervisor/installationgroup.go

```diff
@@ -0,0 +1,183 @@
+			Warn("Another provisioner has worked on this installationGroup; skipping...")
+		return
+	}
+
+	logger.Debugf("Supervising installation group in state %s", installationGroup.State)
+
+	newState := s.transitionInstallationGroup(installationGroup, logger)
+
+	installationGroup, err = s.store.GetInstallationGroupByID(installationGroup.ID)
+	if err != nil {
+		logger.WithError(err).Warnf("failed to get installation group and thus persist state %s", newState)
+		return
+	}
+
+	if installationGroup.State == newState {
+		return
+	}
+
+	oldState := installationGroup.State
+	installationGroup.State = newState
+	if oldState == model.InstallationGroupReleaseSoakingRequested && newState == model.InstallationGroupStable {
+		installationGroup.ReleaseAt = time.Now().UnixNano()
+	}
+
+	if err = s.store.UpdateInstallationGroup(installationGroup); err != nil {
+		logger.WithError(err).Warnf("failed to set installation group state to %s", newState)
+		return
+	}
+
+	webhookPayload := &model.WebhookPayload{
+		Type:      model.TypeRing,
+		ID:        installationGroup.ID,
+		NewState:  newState,
+		OldState:  oldState,
+		Timestamp: time.Now().UnixNano(),
+	}
+	if err = webhook.SendToAllWebhooks(s.store, webhookPayload, logger.WithField("webhookEvent", webhookPayload.NewState)); err != nil {
+		logger.WithError(err).Error("Unable to process and send webhooks")
+	}
+
+	logger.Debugf("Transitioned installation group from %s to %s", oldState, newState)
+}
+
+// Do works with the given ring to transition it to a final state.
+func (s *InstallationGroupSupervisor) transitionInstallationGroup(installationGroup *model.InstallationGroup, logger log.FieldLogger) string {
+	switch installationGroup.State {
+	case model.InstallationGroupReleasePending:
+		return s.checkInstallationGroupPending(installationGroup, logger)
+	case model.InstallationGroupReleaseRequested:
+		return s.releaseInstallationGroup(installationGroup, logger)
+	case model.InstallationGroupReleaseSoakingRequested:
+		return s.soakInstallationGroup(installationGroup, logger)
+	default:
+		logger.Warnf("Found installation group pending work in unexpected state %s", installationGroup.State)
+		return installationGroup.State
+	}
+}
+
+func (s *InstallationGroupSupervisor) checkInstallationGroupPending(installationGroup *model.InstallationGroup, logger log.FieldLogger) string {
+	installationGroups, err := s.store.GetInstallationGroupsPendingWork()
+	if err != nil {
+		logger.WithError(err).Error("Failed to query for installation groups pending work")
+		return model.InstallationGroupReleaseFailed
+	}
+
+	logger.Debug("Checking if other Installations Groups are being already updated...")
+	if len(installationGroups) > 0 {
+		for _, ig := range installationGroups {
+			if ig.State != model.InstallationGroupReleasePending {
+				logger.Debug("Another installation group is being updated...")
+				return model.InstallationGroupReleasePending
+			}
+		}
+	}
+	return model.InstallationGroupReleaseRequested
+}
```

Comment:
If the intention here is so that only one Installation Group is being released at the time, there are possible race conditions.

If multiple servers run line 146 then both of them will conclude that it is safe to release Installation Group. Also given the fact that there is no locking mechanism on the Installation group one server can set it to `InstallationGroupReleaseRequested` while the other will override it to `InstallationGroupReleasePending` or some other status.

I think that if we are going to run multiple instances of Elrond we need to be careful with this as it may lead to hard to debug issues.

Reply:
Yeah you are right on that now that I am thinking about it more clearly. If we move to multi instance setup we will definitely face race conditions. Do you suggest adding lock mechanism for installation groups?

Reply:
For this particular case the most fitting thing that comes to my mind would be to just do the querying and update in single database transaction. This however would not fit supervisor model well as you would need to update the state in this transaction instead of `Supervise` methods as our usual pattern follows.

If we want to follow the Supervisor pattern here, then probably locking is a way to go, I am not sure however what would be the full scope/responsibilities of the Installation Group Supervisor so it is hard for me to confidently say what type of implementation would fit it best.

Reply:
@Szymongib added a locking mechanism and additional check to make sure that no more than one installation group will be released at the same time.